service/test/ConstConfigIOGroupTest.cpp

```diff
@@ -0,0 +1,658 @@
+
+    GEOPM_EXPECT_THROW_MESSAGE(
+        ConstConfigIOGroup iogroup(M_CONFIG_FILE_PATH, ""),
+        GEOPM_ERROR_INVALID,
+        "ConstConfigIOGroup::parse_config_json():"
+        " incorrect type for property: "
+        "\"values\""
+    );
+}
+
+TEST_F(ConstConfigIOGroupTest, input_array_value_empty)
+{
+    std::string json_string = "{"
+    "    \"GPU_CORE_FREQUENCY\": {"
+    "        \"domain\": \"gpu\","
+    "        \"description\": \"Provides GPU core frequency\","
+    "        \"units\": \"hertz\","
+    "        \"aggregation\": \"sum\","
+    "        \"values\": []"
+    "    }"
+    "}";
+    create_config_file(json_string);
+    ConstConfigIOGroup iogroup(M_CONFIG_FILE_PATH, "");
+
+    GEOPM_EXPECT_THROW_MESSAGE(
+        iogroup.read_signal("CONST_CONFIG::GPU_CORE_FREQUENCY", GEOPM_DOMAIN_GPU, 0),
+        GEOPM_ERROR_INVALID,
+        "ConstConfigIOGroup::read_signal(): domain_idx "
+        "0" " out of range."
+    );
+}
+
+TEST_F(ConstConfigIOGroupTest, valid_json_positive)
+{
+    std::string json_string = "{"
+    "    \"GPU_CORE_FREQUENCY\": {"
+    "        \"domain\": \"gpu\","
+    "        \"description\": \"Provides GPU core frequency\","
+    "        \"units\": \"hertz\","
+    "        \"aggregation\": \"sum\","
+    "        \"values\": [ 1500, 1600, 1700 ]"
+    "    },"
+    "    \"CPU_CORE_FREQUENCY\": {"
+    "        \"domain\": \"cpu\","
+    "        \"description\": \"Provides CPU core frequency\","
+    "        \"units\": \"watts\","
+    "        \"aggregation\": \"average\","
+    "        \"values\": [ 1050, 1060, 1070 ]"
+    "    }"
+    "}";
+    create_config_file(json_string);
+    ConstConfigIOGroup iogroup(M_CONFIG_FILE_PATH, "");
+
+    EXPECT_EQ(iogroup.signal_names(), (std::set<std::string>{
+        "CONST_CONFIG::CPU_CORE_FREQUENCY", "CONST_CONFIG::GPU_CORE_FREQUENCY"}));
+    EXPECT_EQ(iogroup.is_valid_signal("CONST_CONFIG::CPU_CORE_FREQUENCY"), true);
+    EXPECT_EQ(iogroup.is_valid_signal("CONST_CONFIG::GPU_CORE_FREQUENCY"), true);
+    EXPECT_EQ(iogroup.signal_domain_type("CONST_CONFIG::CPU_CORE_FREQUENCY"), GEOPM_DOMAIN_CPU);
+    EXPECT_EQ(iogroup.signal_domain_type("CONST_CONFIG::GPU_CORE_FREQUENCY"), GEOPM_DOMAIN_GPU);
+    EXPECT_EQ(iogroup.push_signal("CONST_CONFIG::CPU_CORE_FREQUENCY", GEOPM_DOMAIN_CPU, 1), 0);
+    EXPECT_EQ(iogroup.push_signal("CONST_CONFIG::GPU_CORE_FREQUENCY", GEOPM_DOMAIN_GPU, 2), 1);
+    EXPECT_EQ(iogroup.sample(0), 1060);
+    EXPECT_EQ(iogroup.sample(1), 1700);
+    EXPECT_EQ(iogroup.read_signal("CONST_CONFIG::CPU_CORE_FREQUENCY", GEOPM_DOMAIN_CPU, 2), 1070);
+    EXPECT_EQ(iogroup.read_signal("CONST_CONFIG::GPU_CORE_FREQUENCY", GEOPM_DOMAIN_GPU, 0), 1500);
+    {
+        auto expected_function = geopm::Agg::average;
+        auto actual_function = *iogroup.agg_function("CONST_CONFIG::CPU_CORE_FREQUENCY")
+            .target<double (*) (const std::vector<double>&)>();
+        EXPECT_EQ(expected_function, actual_function);
+    }
```

Comment:
Take a look at ```service/test/geopm_test.hpp```.  It has helper functions for this purpose so you can write this in the form:
```
std::function<double(const std::vector<double> &)> func = iogroup.agg_function("CONST:CONFIG::CPU_CORE_FREQUENCY");
EXPECT_TRUE(is_agg_average(func));
```

Please write these all this way.

Reply:
Ah, cool. Thanks!